### ISSUE TITLE
Add support for automatic mask handling in MultiHeadAttention layer

### ISSUE THREAD

**Issue by ageron** [keras-team/keras#16248]
**System information**.

TensorFlow version (you are using): 2.8.0
Are you willing to contribute it (Yes/No) : Yes, with a bit of guidance

**Describe the feature and the current behavior/state**.

The `MultiHeadAttention` layer (MHA) should support automatic masking, as described in [Understanding masking & padding](https://keras.io/guides/understanding_masking_and_padding/)). Currently, the only masking option is via the `attention_mask` argument, which is powerful, but hard to use.

Benefits:
* Simpler, more concise, less error-prone code for the most common cases, such as self-attention.
* Consistency with other attention layers (`Attention` and `AdditiveAttention`) which _do_ support automatic masking.
* Simpler downstream libraries such as Keras-NLP.
* Simpler code examples for Transformers, which may attract even more users! 😊

To build a Transformer like the one in the [original paper](https://arxiv.org/abs/1706.03762), we need a self-attention layer in the encoder. To implement it, you currently need to explicitly pass the `attention_mask`. It must be computed from the encoder's input padding mask, but you must not forget to add a second axis, to get the shape [B, 1, S], where B is the batch size and S is the maximum length of the input sequences:

```python
import tensorflow as tf
from tensorflow.keras import layers

encoder_input_ids = layers.Input(shape=[None], dtype=tf.int32)
encoder_embeddings = layers.Embedding(1000, 512)(encoder_input_ids)
attn_layer = layers.MultiHeadAttention(num_heads=8, key_dim=512)
encoder_pad_mask = tf.math.not_equal(encoder_input_ids, 0)  # shape [B, S]
attention_mask = encoder_pad_mask[:, tf.newaxis]  # shape [B, 1, S]
outputs = attn_layer(encoder_embeddings, value=encoder_embeddings,
                     attention_mask=attention_mask)
```

I propose that in case the `value` argument carries an automatic mask (`_keras_mask`), then this mask should be used to update the attention mask (something like `new_attention_mask = keras_mask[:, tf.newaxis] & attention_mask`). The code above would simplify to this:

```python
encoder_input_ids = layers.Input(shape=[None], dtype=tf.int32)
encoder_embeddings = layers.Embedding(1000, 512, mask_zero=True)(encoder_input_ids)
attn_layer = layers.MultiHeadAttention(num_heads=8, key_dim=512)
outputs = attn_layer(encoder_embeddings, value=encoder_embeddings)
```

If both the `key` and the `value` are passed to the MHA layer, then their automatic masks (if any) should be merged with a logical AND (something like `pad_mask = value_pad_mask & key_pad_mask`).

The MHA layer should also have `supports_masking` set to `True`, and if the inputs (i.e., the query) have an automatic mask, then that mask should be propagated to the next layer.

Lastly, a new `causal` argument (defaulting to `False`) should be added. This would greatly simplify implementing a masked MHA layer. For example, currently this is the simplest way I could think of to implement a masked MHA:

```python
decoder_input_ids = layers.Input(shape=[None], dtype=tf.int32)
decoder_embeddings = layers.Embedding(1000, 512)(decoder_input_ids)
causal_attn_layer = layers.MultiHeadAttention(num_heads=8, key_dim=512)
decoder_pad_mask = tf.math.not_equal(decoder_input_ids, 0)  # shape [B, T]
T = tf.shape(decoder_input_ids)[1]  # max length of the decoder inputs
causal_mask = tf.linalg.band_part(  # creates a lower triangular matrix
    tf.ones((T, T), tf.bool), -1, 0)
attn_mask = causal_mask & decoder_pad_mask[:, tf.newaxis]
outputs = causal_attn_layer(decoder_embeddings, value=decoder_embeddings,
                     attention_mask=attn_mask)
```

Instead, this proposal would lead to the following code:

```python
decoder_input_ids = layers.Input(shape=[None], dtype=tf.int32)
decoder_embeddings = layers.Embedding(1000, 512, mask_zero=True)(decoder_input_ids)
causal_attn_layer = layers.MultiHeadAttention(num_heads=8, key_dim=512, causal=True)
outputs = causal_attn_layer(decoder_embeddings, value=decoder_embeddings)
```

Notes:
* perhaps the addition of a `causal` argument should be a separate feature request, but since both feature requests would affect the exact same code, I think it would be simpler to treat them jointly, if indeed we want to do both.
* <strike>I believe the `causal` argument should be a constructor argument, but I guess a case could be made that it should be an argument of the `call()` method instead.</strike> **Edit**: as pointed out by @haifeng-jin, the `Attention` layer has a similar `causal` argument, so we should be consistent with it.

**Will this change the current api? How?**

Yes: (1) automatic masks are no longer ignored, and (2) a `causal` argument is added to the constructor.

I don't expect these changes to break existing code, but I think of two ways it could happen:

* If some existing code passes masked inputs to an MHA layer, the masks are currently ignored, and they no longer will be. I cannot see why you would want to ignore the masks, so the change may actually fix a broken model. But perhaps there's a use case I'm not thinking about.
* a subclass of the MHA layer may exist with its own `causal` argument. Seems unlikely, but perhaps it exists.

**Who will benefit from this feature?**

Anyone who want to use the MHA layer with masking.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): yes, with a bit of guidance
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):
I suppose I would edit the MHA layer and implement the masking logic I described above. I'd need a bit of guidance on the internals of masking.

**Comment by haifeng-jin** [comment]
Note: The naming of `causal` is from the argument with same name in the [`Attention`](https://keras.io/api/layers/attention_layers/attention/) layer.

[A tutorial to learn more about Transformer](http://jalammar.github.io/illustrated-transformer/)

**Comment by ageron** [comment]
Ah, good point, thanks. Using `causal` in the constructor makes sense, then. 👍

**Comment by mattdangerw** [comment]
Sorry about the delay here! @ageron are you still interested in working on this? I could help guide (particularly with respect to generating API symbols and helping test this change for any unexpected regressions).

I think the writeup you have looks good, are we actual just did similar work for the keras nlp project. See, for example...
https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/transformer_encoder.py#L159
https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/transformer_layer_utils.py#L38

A few notes...

 - We can probably add a call time argument call `padding_mask` to manually set the padding mask. This would be the exact same behavior as with the `_keras_mask` (combined with `attention_mask`) as described above, but explicit. If both the `_keras_mask` and `padding_mask` are set, we should file a warning, and ignore the `_keras_mask`.
 - We want to make sure to propagate the keras mask (set self.supports_masking = True) on the layer.
 - Re specifying `causal` at init time vs call time, I think init is more probably friendly, unless we think there is a use case to use the same learned weights both with and without causal masking. Nothing pops to mind there.

Anyway thanks so much for the writeup here, this is a very well thought out design!

**Comment by ageron** [comment]
Hi @mattdangerw ,
Thanks for your feedback and your kind words. Yes, I'd be happy to contribute this. I'll try to propose something within the next 2-3 weeks. Is this schedule ok?

**Comment by mattdangerw** [comment]
@ageron works for us! Thank you! Tensorflow 2.9 is about to go live, so we would probably target 2.10 with this. That should give us some lead time.

**Comment by ageron** [comment]
Hi @mattdangerw ,

I've given this a quick shot, the code examples I proposed above work well, but it might need some more work. Could you please take a look at PR #16619?

Some comments regarding your suggestions:

> We can probably add a call time argument call `padding_mask` to manually set the padding mask. This would be the exact same behavior as with the `_keras_mask` (combined with `attention_mask`) as described above, but explicit. If both the `_keras_mask` and `padding_mask` are set, we should file a warning, and ignore the `_keras_mask`.

We may need padding for the `query` and a different padding for the `value` (for cross-attention), so I wasn't sure what to do. Perhaps this can be done in a separate PR, it seems somewhat independent.

> We want to make sure to propagate the keras mask (set `self.supports_masking = True`) on the layer.

Yes, done.

> Re specifying causal at init time vs call time, I think init is probably more friendly, unless we think there is a use case to use the same learned weights both with and without causal masking. Nothing pops to mind there.

I agree. I implemented this at init time, with a new `causal` hyperparameter (defaults to `False`).

Note: to avoid breaking existing code, when both Keras mask(s) and `attention_mask` are defined, they are merged together with a logical AND, currently without a warning. I think this is the safest route for existing code, and it also preserves maximum flexibility, if some users want to handle some of the masking manually, but not all of it (for example if they want to implement some custom padding masking using `attention_mask` for the `value`, but they also want to use Keras masking for the `query`). Wdyt?

**Comment by mattdangerw** [comment]
Thanks for the PR @ageron! Looks great. I will try to do some wider scale internal testing with it now to see if there's any backwards compatibility issues we missed.

> We may need padding for the query and a different padding for the value (for cross-attention), so I wasn't sure what to do. Perhaps this can be done in a separate PR, it seems somewhat independent.

Yeah, that makes sense, and a follow up PR sounds fine to me. But we should probably make sure we know what we want here. Seems like two things we could do (maybe others I am missing)...

1) Offer explicit versions of each of those implicit masks. So optional arguments `query_mask`, `key_mask`, and `value_mask`. These behave exactly the same as the implicit masks in the PR you just put up. If an explicit and implicit mask are provided for the same call argument, log a warning, and take the explicit one.
2) Stick with exactly what you have here. The only way to provide query, key and value masks is through an implicit `_keras_mask` property, and anything else will need an `attention_mask` to be computed manually.

@fchollet any thoughts on this? We may want to follow whatever we decide here in KerasNLP as well.

Re the logical and with the current `attention_mask`, that makes sense to me.

**Comment by mattdangerw** [comment]
Checked in with @fchollet about this. His take was we should do option 1) above, this won't add clutter for the default usage and will follow other layers in terms of masking ux, where an implicit mask can always be given explicitly as well.

Happy to take that on in a follow up PR though, we don't need to add that to this change.

**Comment by ageron** [comment]
Sounds good to me. As soon as this PR is merged, I'm happy to help with adding the new arguments.

**Comment by mattdangerw** [comment]
@ageron sounds good! You are welcome to do the follow up as well. Just didn't want to rope you in without checking :)

**Comment by mattdangerw** [comment]
OK! Finally landed this.

@ageron if you have time, would you like to open a PR for adding the explicit mask arguments? Thank you!

**Comment by saberkun** [comment]
is it OK to rely on _keras_mask?

Why don't we pass the causal mask as the attention mask to keep the same flexibility and reduce assumptions?
Note, if we pass attention mask, it will not calculate the mask operation per layer. If we have 100+ layers, the mask is the same and shared across layers.

**Comment by fchollet** [comment]
With this API you still have the option to specify the mask manually via the `attention_mask` argument as before.

This change simply makes it easier to implement simpler workflows that rely on the masks automatically generated by the `Embedding` layer.
